contracts/Registry.sol

```diff
@@ -0,0 +1,182 @@
+
+struct VaultData {
+    VaultStatus status;
+    address vaultAddress;
+    string vaultName;
+    string vaultSymbol;
+    string vaultApiVersion;
+    address tokenAddress;
+    string tokenName;
+    string tokenSymbol;
+    uint256 tokenDecimals; // NOTE: Vault.decimals() == Token.decimals()
+}
+
+contract Registry {
+    using Address for address;
+    using SafeMath for uint256;
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    address public governance;
+    address pendingGovernance;
+
+    EnumerableSet.AddressSet private vaults;
+    mapping(address => address) public latestVaultForToken;
+
+    // Keep track of status
+    EnumerableSet.AddressSet private experimentalVaults;
+    EnumerableSet.AddressSet private endorsedVaults;
+    EnumerableSet.AddressSet private deprecatedVaults;
+
+    event VaultUpgrade(address indexed token, address indexed oldVault, address newVault);
+
+    constructor(address _governance) public {
+        require(_governance != address(0), "Missing Governance");
+        governance = _governance;
+    }
+
+    // Governance setters
+    function setGovernance(address _governance) external onlyGovernance {
+        pendingGovernance = _governance;
+    }
+
+    function acceptGovernance() external {
+        require(msg.sender == pendingGovernance, "!pendingGovernance");
+        governance = msg.sender;
+    }
+
+    modifier onlyGovernance {
+        require(msg.sender == governance, "!governance");
+        _;
+    }
+
+    // Vault set actions
+    function addVault(address _vault, bool experimental) public onlyGovernance {
+        require(_vault.isContract(), "Vault is not a contract");
+        // Checks if vault is already on the array
+        require(!vaults.contains(_vault), "Vault already exists");
+        // Adds unique _vault to vaults array
+        vaults.add(_vault);
+
+        if (experimental) {
+            experimentalVaults.add(_vault);
+        } else {
+            endorsedVaults.add(_vault);
+        }
+
+        // Keep track of latest Vault (newest is considered best)
+        // NOTE: A vault should only be added to the registry when it's considered
+        //       "production" ready, signaling it as the one people should upgrade to
+        address token = VaultAPI(_vault).token();
+        address oldVault = latestVaultForToken[token];
+        latestVaultForToken[token] = _vault;
+        emit VaultUpgrade(token, oldVault, _vault);
+    }
+
+    function endorseVault(address _vault) public onlyGovernance {
+        require(experimentalVaults.contains(_vault), "!experimental");
+        experimentalVaults.remove(_vault);
+        endorsedVaults.add(_vault);
+    }
+
+    function removeVault(address _vault) public onlyGovernance {
+        // Checks if vault is already on the array
+        require(vaults.contains(_vault), "Vault not in set");
+        // Remove _vault to vaults array
+        vaults.remove(_vault);
```

Comment:
Do you need to have the vaults in two different lists? Maybe the status can be done using a mapping, so then you only need one list for the vault. 

```mapping(address => VaultStatus) public  status;```

Reply:
I wanted to add a status enum to each Vault, but the end goal here is to make it easy for UI integration. From a UI perspective, it wants to render the list of Vaults, depending on what the user cases about.

A user cares about a few things:
1. Do I need to "upgrade" any of my Vault shares (`getDeprecatedVaults()` migrate to `getEndorsedVaults()`)
2. Do I want to use any of the Vaults in "beta" (`getExperimentalVaults()`)
3. What are my current holdings (`getActiveVaults()`)

Being able to view them in a list format works better for JS, otherwise we'd be maintaining an Enum in two separate places. We'd also have to perform an interation over a series of calls that's avoided from having these lists. It's definitely not the most efficient thing in the world, but there's some method to the madness.